app/models/test_track/session.rb

```diff
@@ -31,23 +31,41 @@ def state_hash
     }
   end
 
-  def log_in!(identifier_type, identifier_value, opts = {})
+  def log_in!(identity, opts = {})
     @visitor = TestTrack::Visitor.new if opts[:forget_current_visitor]
-    visitor.link_identifier!(identifier_type, identifier_value)
+    visitor.link_identifier!(identity.test_track_identifier_type, identity.test_track_identifier_value)
+    managed_identities[identity.test_track_identifier_type] = identity
     self.mixpanel_distinct_id = visitor.id
     true
   end
 
-  def sign_up!(identifier_type, identifier_value)
-    visitor.link_identifier!(identifier_type, identifier_value)
+  def sign_up!(identity)
+    visitor.link_identifier!(identity.test_track_identifier_type, identity.test_track_identifier_value)
+    managed_identities[identity.test_track_identifier_type] = identity
     @signed_up = true
   end
 
+  def managed_identity?(identity)
+    managed_identity = managed_identities[identity.test_track_identifier_type] ||= authenticated_resource_for_identity(identity)
+    managed_identity.present? && managed_identity == identity
```

Comment:
a lot going on in these two lines. This feels a little bit like object envy maybe? Like this might be easier to grok if `IdentityCollection`were another class or something? like:

```ruby
class TestTrack::Session
  ...

  def managed_identities
    @managed_identities = SessionIdentityCollection.new(controller)
  end

  ...
end

class SessionIdentityCollection
  def initialize(controller)
    ...
  end

  def find_by_identifier_type(identifier_type)
    ...
  end

  def <<(identity)
    ...
  end
end
```

Reply:
Is `managed_identity?` only going to return false if the controller-authenticated identity doesn't match the requested identity? Maybe we should name the methods for getting at this fact to reflect that authorization concern? Because if I'm reading correctly it'll lazily start managing any identity not in conflict with the controller identity and return true, so `managed` doesn't sound right?

Reply:
`Session#has_matching_identity?(identity)` or something?